### ISSUE TITLE
investigate browser support

### ISSUE THREAD

**Issue by chase-moskal** [chase-moskal/webp-hero#8]
let's test the [live demo in force mode](https://chasemoskal.com/webp-hero/?force) in various browsers, see if the images decode correctly, and fix anything that's broken!

desktop
- [x] :heavy_check_mark: *good* — chrome 71
- [x] :heavy_check_mark: *good* — firefox
- [x] :heavy_check_mark: *good* — ie 11 — ~~#11~~ fixed :dancer:
- [x] :heavy_check_mark: *good* — edge 42 (edgehtml 17) — thanks @06Games
- [ ] ??? — safari

mobile
- [x] :heavy_check_mark: *good* — chrome
- [x] :heavy_check_mark: *good* — ios safari
- [ ] ??? — firefox

**Comment by achriwiro** [comment]
Tried your testing URL in IE11 with '?force' and got this error:
`SCRIPT1003: Expected ':'
const webpMachine = new webpHero.WebpMachine({webpSupport})`
: is expected after 'webpSupport' on the above line. Though I'm not sure why that would be the case.

**Comment by chase-moskal** [comment]
@Lumbearjack — thank you for testing it in IE11 and letting me know, much appreciated!

i've added an npm script "polyfills" to `package.json` which generates `webp-hero/dist/polyfills.js`, restores support to IE11

- the webp-hero consumer must opt-in to using `polyfills.js`, it's not included in `webp-hero.bundle.js`
- the fix is merged with commit 6334019
- the fix has been published to npm `webp-hero@0.0.0-dev.18`
- the fix has been deployed to the live demo https://chasemoskal.com/webp-hero/
- [changelog](https://github.com/chase-moskal/webp-hero/blob/master/CHANGELOG.md#v000-dev18--2019-02-26) has been updated
- [readme](https://github.com/chase-moskal/webp-hero/blob/master/README.md#use-webp-hero-in-a-simple-html-page) instructions have been updated

here are the changes we ended up needing for IE11
- switching from `const` to `var` on the demo page
- `es6-promise`
- `mdn-polyfills` / Array.from.js

let me know if you notice any more problems :beers:

**Comment by jnorthrup** [comment]
the experience of running in nextjs app/server is that there is no satisfying the initial check for "document"

the code runs in the browser statically as expected.

testing on an iphone, the page crashed in safari and chrome, likely OOM.  howeever the developer im working with was able to manually integrate image-by-image react rendering for exported static content.  

due to the presumed OOM, we just bit the bullet and added a jpg asset alongside the webp and the developer chose his favorite image-tag-workaround library.

**Comment by jnorthrup** [comment]
fwiw, firefox 63 is available for download and excludes webp support.

**Comment by chase-moskal** [comment]
confirmed it works on ios safari, tested with an ipad the other day

**Comment by 06Games** [comment]
It works fine with **EdgeHTML 17** (Edge 42).
Thank you for your library.

**Comment by chase-moskal** [comment]
thanks @06Games, i've updated the list :+1: :+1:

**Comment by MacedoRAC** [comment]
First of all , @chase-moskal , thank you for this awesome package.

Regarding this situation of IE11, I'm using version v0.0.0-dev.24 of your package and I'm importing the "webp-hero/dist/polyfills.js" in my project but I'm still facing an issue with Async/Await. For some reason some functions aren't being transpiled. 

```javascript
export async function detectWebpSupport() {
    const testImageSources = [
        "data:image/webp;base64,UklGRjIAAABXRUJQVlA4ICYAAACyAgCdASoCAAEALmk0mk0iIiIiIgBoSygABc6zbAAA/v56QAAAAA==",
        "data:image/webp;base64,UklGRh4AAABXRUJQVlA4TBEAAAAvAQAAAAfQ//73v/+BiOh/AAA="
    ];
    const testImage = (src) => {
        return new Promise((resolve, reject) => {
            var img = document.createElement("img");
            img.onerror = error => resolve(false);
            img.onload = () => resolve(true);
            img.src = src;
        });
    };
    const results = await Promise.all(testImageSources.map(testImage));
    return results.every(result => !!result);
}
```

I'm using your package in a VueCli 4 project and below you can see my main entry point of the project where I'm  importing the required files for the package to work.

```javascript
import Vue from 'vue'
import App from './App.vue'
import router from './router'

import './assets/scss/app.scss'

require ('webp-hero/dist-cjs/polyfills.js')
import {WebpMachine} from "webp-hero"
const webpMachine = new WebpMachine()
webpMachine.polyfillDocument()

Vue.config.productionTip = false

new Vue({
  router,
  render: h => h(App)
}).$mount('#app')
```

In attachment you can find the package.json, babel.config.js and .browserslistrc, but it's a standard configuration of a VueCli project

[files.zip](https://github.com/chase-moskal/webp-hero/files/4509788/files.zip)

**Comment by chase-moskal** [comment]
@MacedoRAC sorry for my late reply

`webp-hero/dist/**` — target `esnext` — cutting edge es modules distribution
`webp-hero/dist-cjs/**` — target `es5` — legacy-compatible commonjs modules
`webp-hero/dist-cjs/webp-hero.bundle.js` — browserify bundle, attaches `webpHero` global object to window

> I'm still facing an issue with Async/Await. For some reason some functions aren't being transpiled.

see
- [dist/detect-webp-support.js](https://unpkg.com/browse/webp-hero@0.0.0-dev.24/dist/detect-webp-support.js) (modern esm)
- [dist-cjs/detect-webp-support.js](https://unpkg.com/browse/webp-hero@0.0.0-dev.24/dist-cjs/detect-webp-support.js) (transpiled for compatibility)

**Comment by chase-moskal** [comment]
@MacedoRAC — i think you need to import from `webp-hero/dist-cjs/` explicitly because your environment for some reason is defaulting to the es module distribution, not sure

**Comment by MacedoRAC** [comment]
@chase-moskal thank you very much for your help. 

changing the import from

```
import {WebpMachine} from "webp-hero"
```
to

```
import {WebpMachine} from "webp-hero/dist-cjs"
```

actually did the trick.

**Comment by Macleykun** [comment]
Hello, is there anything I can do to help out?
I wanted to let the dev know that safari is actually getting webp support
https://developer.apple.com/documentation/safari-release-notes/safari-14-beta-release-notes
Might be handy to adjust the code for this change :)

**Comment by chase-moskal** [comment]
@Macley-Kun hello! i have heard about safari introducing webp support, that's great

at this time, we're blocked on upgrading webp-hero until we can get the google libwebp build working, i got stuck on that task and had to abandon it to focus on other things...

we have some other open issues about it

but the plan generally, if i ever get around to it, is to

1. fix the google libwebp build (happens inside a docker container here in the webp-hero repo)
2. consider some refactors and incorporating features and fixes from various open issues
3. test all browsers with a tool like browserstack
4. publish a new stable version of webp-hero

it's possible that next month i might have some time to spare to dive into this, but if anybody's willing to contribute, the best i can do is provide pointers on discord or something on how to proceed

it's been awhile since i've attempted the libwebp build, so that problem may resolve itself if we try again with fresh versions of libwebp and emscripten

**Comment by Macleykun** [comment]
Sure np! I hope libwebp will work for ya in docker and that you can work on this wonderful project! Gl on it!

**Comment by chase-moskal** [comment]
hello! i'm closing this issue as browser support is now mentioned in the readme

if you experience any issues in any browsers that matter to you, please open a new issue!

thanks everyone involved!
